File: packages/chat-sync/src/chat-history.ts
import type { FlueConversationMessage } from '@flue/sdk';

export type StoredChatMessage = {
  id: string;
  contentJson: string;
  createdAt: number;
  hidden?: boolean;
};

export type ChatMessageSyncEnvelope = {
  id: string;
  role: 'user' | 'assistant';
  contentJson: string;
  createdAt: number;
};

const LIVE_SYNC_TAIL_SIZE = 8;

function messageKeys(message: FlueConversationMessage) {
  return [
    `id:${message.id}`,
    ...(message.role === 'user' && message.submissionId
      ? [`submission:${message.submissionId}`]
      : []),
  ];
}

/** True only for primitive strings; boxing avoids invoking value `toString`s. */
function isStringValue<Value>(value: Value): value is Value & string {
  return Object(value) instanceof String;
}

function messageTimestamp(message: FlueConversationMessage, fallback: number) {
  // Agent-authored metadata carries `timestamp` as an ISO-8601 string by
  // convention; anything else falls back to the caller-supplied ordering.
  const value = message.metadata?.timestamp;
  const parsed = isStringValue(value) ? Date.parse(value) : Number.NaN;
  return Number.isFinite(parsed) ? parsed : fallback;
}

function findMessagePosition(
  entry: { message: FlueConversationMessage; createdAt: number },
  position: Map<string, number>,
) {
  return messageKeys(entry.message)
    .map((key) => position.get(key))
    .find((value) => value !== undefined);
}

export type SyncableMessage = FlueConversationMessage & {
  role: 'user' | 'assistant';
};

/** True for the user/assistant turns Convex stores as chat history. */
function isSyncable(message: FlueConversationMessage): message is SyncableMessage {
  return message.role !== 'system';
}

/** Gives admitted user turns a stable key while Flue reconciles its local echo. */
export function messagesForConvexSync(
  messages: FlueConversationMessage[],
): SyncableMessage[] {
  return messages.flatMap((message) => {
    // Flue 2.0 folds dispatch signals and runtime advisories into
    // system-role messages; they are runtime plumbing, not chat history.
    if (!isSyncable(message)) {
      return [];
    }
    if (message.role === 'user' && message.submissionId) {
      return [{ ...message, id: `submission:${message.submissionId}` }];
    }
    if (message.id.startsWith('local:')) {
      return [];
    }
    return [message];
  });
}

/** Serializes only unknown messages and the active tail of a live snapshot. */
export function changedMessagesForConvexSync(
  messages: FlueConversationMessage[],
  knownContent: ReadonlyMap<string, string>,
): ChatMessageSyncEnvelope[] {
  const syncable = messagesForConvexSync(messages);
  const tailStart = Math.max(0, syncable.length - LIVE_SYNC_TAIL_SIZE);

  return syncable.flatMap((message, index) => {
    const previous = knownContent.get(message.id);
    if (previous !== undefined && index < tailStart) return [];

    const contentJson = JSON.stringify(message);
    if (previous === contentJson) return [];

    return [{
      id: message.id,
      role: message.role,
      contentJson,
      createdAt: messageTimestamp(message, index),
    }];
  });
}

/** True when a merged message matches a tombstoned Convex row. */
function isHiddenMessage(
  message: FlueConversationMessage,
  hiddenIds: ReadonlySet<string>,
) {
  return (
    hiddenIds.has(message.id) ||
    (message.role === 'user' &&
      message.submissionId !== undefined &&
      hiddenIds.has(`submission:${message.submissionId}`))
  );
}

/** Combines the durable transcript with Flue's live streaming envelope. */
export function mergeConvexMessages(
  rows: StoredChatMessage[] | undefined,
  flueMessages: FlueConversationMessage[],
) {
  const ordered: { message: FlueConversationMessage; createdAt: number }[] = [];
  const position = new Map<string, number>();

  for (const row of rows ?? []) {
    try {
      // SAFETY: stored rows are written only from changedMessagesForConvexSync,
      // which serializes FlueConversationMessage values verbatim, so a parsed
      // row is one of our own messages; a corrupted row throws here and is
      // skipped by the catch below.
      const message = JSON.parse(row.contentJson) as FlueConversationMessage;
      const entry = {
        message,
        createdAt: messageTimestamp(message, row.createdAt),
      };
      const existing = findMessagePosition(entry, position);
      if (existing === undefined) {
        const nextIndex = ordered.length;
        ordered.push(entry);
        for (const key of messageKeys(message)) position.set(key, nextIndex);
      } else {
        const previous = ordered[existing];
        ordered[existing] = {
          message,
          createdAt: Math.min(previous.createdAt, entry.createdAt),
        };
        for (const key of [
          ...messageKeys(previous.message),
          ...messageKeys(message),
        ]) {
          position.set(key, existing);
        }
      }
    } catch {
      // One malformed stored envelope should not hide the rest of the chat.
    }
  }

  const fallbackTimestamp =
    ordered.reduce((latest, entry) => Math.max(latest, entry.createdAt), 0) + 1;

  for (const [index, message] of flueMessages.entries()) {
    const entry = {
      message,
      createdAt: messageTimestamp(message, fallbackTimestamp + index),
    };
    const existing = findMessagePosition(entry, position);
    if (existing === undefined) {
      const nextIndex = ordered.length;
      for (const key of messageKeys(message)) position.set(key, nextIndex);
      ordered.push(entry);
    } else {
      for (const key of messageKeys(message)) position.set(key, existing);
      ordered[existing] = { ...ordered[existing], message };
    }
  }

  const hiddenIds = new Set(
    (rows ?? []).filter((row) => row.hidden).map((row) => row.id),
  );

  return ordered
    .sort((left, right) => left.createdAt - right.createdAt)
    .map(({ message }) => message)
    .filter((message) => !isHiddenMessage(message, hiddenIds));
}

File: apps/android/core/contract/src/main/kotlin/com/beegreat/contract/ChatHistory.kt
package com.beegreat.contract

import com.beegreat.flue.FlueJson
import com.beegreat.flue.FlueMessage

// Port of packages/chat-sync/src/chat-history.ts.

data class StoredChatMessage(val id: String, val contentJson: String, val createdAt: Long, val hidden: Boolean = false)

data class ChatMessageSyncEnvelope(val id: String, val role: String, val contentJson: String, val createdAt: Long)

private const val LIVE_SYNC_TAIL_SIZE = 8

private fun messageKeys(message: FlueMessage): List<String> =
  buildList {
    add("id:${message.id}")
    if (message.isUser && message.submissionId != null) add("submission:${message.submissionId}")
  }

private fun messageTimestamp(message: FlueMessage, fallback: Long): Long = message.timestampMillis ?: fallback

/** True for the user/assistant turns Convex stores as chat history. */
private fun isSyncable(message: FlueMessage) = message.role != "system"

/** Gives admitted user turns a stable key while Flue reconciles its local echo. */
fun messagesForConvexSync(messages: List<FlueMessage>): List<FlueMessage> =
  messages.mapNotNull { message ->
    when {
      !isSyncable(message) -> null
      message.isUser && message.submissionId != null -> message.copy(id = "submission:${message.submissionId}")
      message.id.startsWith("local:") -> null
      else -> message
    }
  }

fun encodeMessage(message: FlueMessage): String = FlueJson.encodeToString(FlueMessage.serializer(), message)

/** Serializes only unknown messages and the active tail of a live snapshot. */
fun changedMessagesForConvexSync(messages: List<FlueMessage>, knownContent: Map<String, String>): List<ChatMessageSyncEnvelope> {
  val syncable = messagesForConvexSync(messages)
  val tailStart = maxOf(0, syncable.size - LIVE_SYNC_TAIL_SIZE)
  return syncable.mapIndexedNotNull { index, message ->
    val previous = knownContent[message.id]
    if (previous != null && index < tailStart) return@mapIndexedNotNull null
    val contentJson = encodeMessage(message)
    if (previous == contentJson) return@mapIndexedNotNull null
    ChatMessageSyncEnvelope(message.id, message.role, contentJson, messageTimestamp(message, index.toLong()))
  }
}

private fun isHiddenMessage(message: FlueMessage, hiddenIds: Set<String>) =
  message.id in hiddenIds || (message.isUser && message.submissionId != null && "submission:${message.submissionId}" in hiddenIds)

/** Combines the durable transcript with Flue's live streaming envelope. */
fun mergeConvexMessages(rows: List<StoredChatMessage>?, flueMessages: List<FlueMessage>): List<FlueMessage> {
  class Entry(var message: FlueMessage, var createdAt: Long)
  val ordered = ArrayList<Entry>()
  val position = HashMap<String, Int>()

  fun findPosition(message: FlueMessage) = messageKeys(message).firstNotNullOfOrNull { position[it] }

  for (row in rows ?: emptyList()) {
    val message = runCatching { FlueJson.decodeFromString(FlueMessage.serializer(), row.contentJson) }.getOrNull() ?: continue
    val createdAt = messageTimestamp(message, row.createdAt)
    val existing = findPosition(message)
    if (existing == null) {
      val index = ordered.size
      ordered += Entry(message, createdAt)
      for (key in messageKeys(message)) position[key] = index
    } else {
      val previous = ordered[existing]
      val keys = messageKeys(previous.message) + messageKeys(message)
      previous.message = message
      previous.createdAt = minOf(previous.createdAt, createdAt)
      for (key in keys) position[key] = existing
    }
  }

  val fallbackTimestamp = (ordered.maxOfOrNull { it.createdAt } ?: 0L) + 1
  for ((index, message) in flueMessages.withIndex()) {
    val createdAt = messageTimestamp(message, fallbackTimestamp + index)
    val existing = findPosition(message)
    if (existing == null) {
      val next = ordered.size
      for (key in messageKeys(message)) position[key] = next
      ordered += Entry(message, createdAt)
    } else {
      for (key in messageKeys(message)) position[key] = existing
      ordered[existing].message = message
    }
  }

  val hiddenIds = (rows ?: emptyList()).filter { it.hidden }.map { it.id }.toSet()
  return ordered
    .sortedBy { it.createdAt }
    .map { it.message }
    .filter { isSyncable(it) && it.display != "hidden" && it.display != "diagnostic" }
    .filterNot { isHiddenMessage(it, hiddenIds) }
}
